### PR TITLE
AGENT-104:  Add Dockerfiles for OCP release

### PR DESCRIPTION
**Pull request by pawanpinjarkar**
Ref: [AGENT-104](https://issues.redhat.com//browse/AGENT-104)
- Added `Dockerfile.assisted-service.ocp`
- Removed `vendor/` from .gitgnore
-  Run `go mod vendor` 
-  Add in `vendor` directory

Additionally, 
- Fix reflect vendoring error 
Ref: https://github.com/golang/mock#debugging-errors
- Do not delete the mock files from the vendor directory else unit-tests will fail.

### CODE REVIEW

**Comment by omertuc** [comment]
Can you add the vendor-diff target as a prerequisite to the ci-lint target?

File: Makefile
Comment (omertuc) [inline Makefile:160]:
```suggestion
ci-lint: vendor-diff
```

**Comment by andfasano** [comment]
Hey @pawanpinjarkar @omertuc, anything missing on this PR to proceed?

**Comment by omertuc** [comment]
> Hey @pawanpinjarkar @omertuc, anything missing on this PR to proceed?

The CI should be green

**Comment by andfasano** [comment]
thanks @omertuc I see, especially the `verify-generate-code` one (not sure about the other two).
In general, my personal preference when revendoring would be to keep two separate commits, one for the pure vendoring part, and the other for any eventual impact on the code

**Comment by omertuc** [comment]
>thanks @omertuc I see, especially the verify-generate-code one (not sure about the other two).

Yeah @pawanpinjarkar is doing some efforts to try and resolve these issues. I believe it's still work in progress.

>In general, my personal preference when revendoring would be to keep two separate commits, one for the pure vendoring part, and the other for any eventual impact on the code

I agree, it's nicer for review. But doesn't really matter too much because our merge strategy squashes all the commits anyway.

**Comment by pawanpinjarkar** [comment]
`skipper make generate-all` now runs fine locally.

File: hack/Makefile
Comment (osherdp) [inline hack/Makefile:17]: but it means we're left with redundant mock files whenever we're deleting the mocked modules, no?

**Comment by osherdp** [comment]
/assign @filanov 
you should review this. I'm less familiar with mockgen and vendoring

File: Makefile
Comment (osherdp) [inline Makefile:159]: ``ci-lint`` is rather unused, but you can add vendor generation to CI right in this PR
You can have in ``hack/Makefile``:
```
generate-vendor:
    go mod vendor

generate: generate-vendor generate-from-swagger generate-events generate-mocks generate-configuration
```

File: internal/bminventory/inventory.go
Comment (osherdp) [inline internal/bminventory/inventory.go:130]: why 3 times?

File: internal/controller/controllers/common.go
Comment (osherdp) [inline internal/controller/controllers/common.go:50]: can we add this build_flag more globally to the ``go generate`` command?
e.g. ``go generate --mod=mod ...``

File: internal/manifests/api/interface.go
Comment (osherdp) [inline internal/manifests/api/interface.go:11]: and here why the same argument needs to appear twice?

**Review by osherdp — CHANGES_REQUESTED** [review]

File: Makefile
Comment (omertuc) [reply Makefile:159]: >ci-lint is rather unused

Then where do we validate tehat generation doesn't result in diff?

Comment (osherdp) [reply Makefile:159]: https://github.com/openshift/release/blob/d50e1472bb34baaee6f240ddafc9a0b467d250f5/ci-operator/step-registry/assisted/verify-generated-code/assisted-verify-generated-code-commands.sh#L14

Comment (osherdp) [reply Makefile:159]: Maybe we can do diff of git-diffs, but it's rather ugly

File: hack/Makefile
Comment (pawanpinjarkar) [reply hack/Makefile:17]: If not removed, then `skipper make unit-test` fails with
```
vendor/github.com/ovirt/go-ovirt-client/mock.go:27:48: undefined: diskWithData
vendor/github.com/ovirt/go-ovirt-client/mock.go:34:48: undefined: datacenterWithClusters
vendor/github.com/ovirt/go-ovirt-client/newmock.go:76:18: undefined: datacenterWithClusters
vendor/github.com/ovirt/go-ovirt-client/newmock.go:131:42: undefined: datacenterWithClusters
vendor/github.com/ovirt/go-ovirt-client/newmock.go:139:52: undefined: datacenterWithClusters
```
because when the above line is in, then `skipper make generate` deletes the mock files
```
deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_cluster_get.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_cluster_list.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_datacenter.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_datacenter_get.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_datacenter_list.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_datacenter_list_clusters.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_disk.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_disk_attachment_create.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_disk_attachment_get.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_disk_attachment_list.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_disk_attachment_remove.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_disk_create.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_disk_downloadimage.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_disk_get.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_disk_get_internal.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_disk_list.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_disk_list_by_alias.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_disk_remove.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_disk_update.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_disk_uploadimage.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_disk_util.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_host_get.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_host_list.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_network_get.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_network_list.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_nic_create.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_nic_get.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_nic_list.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_nic_remove.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_nic_update.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_storagedomain_get.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_storagedomain_get_disk.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_storagedomain_list.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_storagedomain_remove_disk.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_tag_create.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_tag_get.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_tag_list.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_tag_remove.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_template_copy_disk.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_template_create.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_template_get.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_template_get_blank.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_template_list.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_template_remove.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_template_wait_for_status.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vm_add_tag.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vm_create.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vm_get.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vm_list.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vm_optimize_pinning.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vm_remove.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vm_search.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vm_shutdown.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vm_start.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vm_stop.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vm_update.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vm_wait_for_status.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vnicprofile_create.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vnicprofile_get.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vnicprofile_list.go
        deleted:    vendor/github.com/ovirt/go-ovirt-client/mock_vnicprofile_remove.go
```

File: internal/manifests/api/interface.go
Comment (pawanpinjarkar) [reply internal/manifests/api/interface.go:11]: typo

File: internal/controller/controllers/common.go
Comment (pawanpinjarkar) [reply internal/controller/controllers/common.go:50]: Not sure. The doc says https://github.com/golang/mock#debugging-errors

File: hack/Makefile
Comment (osherdp) [reply hack/Makefile:17]: ahhh...
so you can change to ``find . -name 'mock_*.go' -type f -not -path "vendor/*" -delete``

File: internal/controller/controllers/common.go
Comment (pawanpinjarkar) [reply internal/controller/controllers/common.go]: Added build_flag globally to the `go generate` command

File: hack/Makefile
Comment (filanov) [reply hack/Makefile:17]: +1 to Osher

File: internal/controller/controllers/crd_events_handler.go
Comment (osherdp) [inline internal/controller/controllers/crd_events_handler.go:10]: nit: seems like a redundant change

**Review by osherdp — CHANGES_REQUESTED** [review]

**Comment by omertuc** [comment]
Can you please rebase and edit/organize your commits so that we have a simpler view of the changes here?

**Comment by filanov** [comment]
How does vendor work with sub-modules? do you need to have a vendor for each sub module?

File: hack/Makefile
Comment (osherdp) [inline hack/Makefile:18]: wait a second
I think ``go generate -run`` isn't really the way to pass those ``build_flags`` arguments. It allows providing a regex for filtering the ``go:generate`` directives
If not possible globally, then we should revert back to having the build_flags for each ``go:generate`` directive

File: hack/Makefile
Comment (osherdp) [reply hack/Makefile:17]: I'm afraid it must be exactly ``find . -name 'mock_*.go' -type f -not -path "vendor/*" -delete``. With latest changes no mock files are generated

File: hack/Makefile
Comment (osherdp) [inline hack/Makefile:55]: no need for the `` && git diff --exit-code vendor``, locally it doesn't really do as intended
CI should handle comparing actual ``git diff``

**Review by osherdp — CHANGES_REQUESTED** [review]
/hold
I think mock files are not generated, so some changes are needed

File: hack/Makefile
Comment (pawanpinjarkar) [reply hack/Makefile]: I don't think there is any other alternative. I am going back to the original changes.

File: hack/Makefile
Comment (pawanpinjarkar) [reply hack/Makefile:17]: Nit: `find . -name 'mock_*.go' -type f -not -path './vendor/*' -delete`

**Comment by pawanpinjarkar** [comment]
@omertuc @osherdp @filanov Please take a look. The commits are now re-written into the smallest self-explanatory subtask.

The only thing to be checked next into the source control is the generated `vendor` directory which I will add as another commit to this PR later once the current changes at this point look ok. At this point, it should be straightforward to review as vendoring changes won't obstruct the code review process. Till then, please take a look at the current changes. 

You can locally do `skipper make generate` and `skipper make unit-test` to see if file generations and unit tests works fine.

**Review by osherdp — APPROVED** [review]

**Review by osherdp — APPROVED** [review]

**Comment by osherdp** [comment]
> How does vendor work with sub-modules? do you need to have a vendor for each sub module?

@pawanpinjarkar can you elaborate on that one?
/hold

**Comment by filanov** [comment]
We have two sub-modules
https://github.com/openshift/assisted-service/tree/master/api
https://github.com/openshift/assisted-service/tree/master/models
Each have `go.mod` file

File: hack/Makefile
Comment (osherdp) [inline hack/Makefile:17]: with ``api/vendor`` and ``models/vendor`` shouldn't it be updated now?

**Review by osherdp — CHANGES_REQUESTED** [review]
@pawanpinjarkar seems like it [fails](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_assisted-service/3660/pull-ci-openshift-assisted-service-master-images/1521408021572358144) to get built:
```
[3/5] STEP 11/13: RUN CGO_ENABLED=1 GOFLAGS="" GO111MODULE=on go build -o /build/assisted-service cmd/main.go
go: inconsistent vendoring in /go/src/github.com/openshift/origin:
	github.com/google/go-cmp@v0.5.8: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring@v0.56.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	gorm.io/gorm@v1.23.5: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/google/go-cmp@v0.5.7: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
	github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring@v0.55.1: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
	gorm.io/gorm@v1.23.4: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
	To ignore the vendor directory, use -mod=readonly or -mod=mod.
	To sync the vendor directory, run:
		go mod vendor
```

File: hack/Makefile
Comment (pawanpinjarkar) [reply hack/Makefile:17]: The mock files were not deleted from `api/vendor` and `models/vendor` hence no need to exclude here

**Comment by pawanpinjarkar** [comment]
> @pawanpinjarkar seems like it [fails](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_assisted-service/3660/pull-ci-openshift-assisted-service-master-images/1521408021572358144) to get built:
> 
> ```
> [3/5] STEP 11/13: RUN CGO_ENABLED=1 GOFLAGS="" GO111MODULE=on go build -o /build/assisted-service cmd/main.go
> go: inconsistent vendoring in /go/src/github.com/openshift/origin:
> 	github.com/google/go-cmp@v0.5.8: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
> 	github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring@v0.56.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
> 	gorm.io/gorm@v1.23.5: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
> 	github.com/google/go-cmp@v0.5.7: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
> 	github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring@v0.55.1: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
> 	gorm.io/gorm@v1.23.4: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
> 	To ignore the vendor directory, use -mod=readonly or -mod=mod.
> 	To sync the vendor directory, run:
> 		go mod vendor
> ```

I tried building the image locally and it got built successfully.
```
docker build -f Dockerfile.assisted-service -t localassisted .
Sending build context to Docker daemon  213.1MB
Step 1/34 : FROM quay.io/edge-infrastructure/swagger-codegen-cli:2.4.18 as swagger_py
 ---> bc1d8d512b34
...
...
Step 19/34 : RUN CGO_ENABLED=1 GOFLAGS="" GO111MODULE=on go build -o /build/assisted-service cmd/main.go
 ---> Running in 3a08ce7bbf68
Removing intermediate container 3a08ce7bbf68
 ---> 99033e69f091
Step 20/34 : RUN CGO_ENABLED=0 GOFLAGS="" GO111MODULE=on go build -o /build/assisted-service-operator cmd/operator/main.go
 ---> Running in 10040389ea52
Removing intermediate container 10040389ea52
 ---> 63d95a720285
Step 21/34 : RUN CGO_ENABLED=0 GOFLAGS="" GO111MODULE=on go build -o /build/assisted-service-admission cmd/webadmission/main.go
 ---> Running in 6067bde0d845
Removing intermediate container 6067bde0d845
 ---> 217dc9ee9bd9
Step 22/34 : FROM quay.io/ocpmetal/oc-image:bug-1823143-multi-arch as oc-image
 ---> 8a3799e50bf1
....
....
Successfully built a39dae7a1129
Successfully tagged localassisted:latest
```

**Comment by pawanpinjarkar** [comment]
Rebased against the master. The CI failures seen above should resolve :crossed_fingers:

**Review by osherdp — APPROVED** [review]

**Pull request merged**
